### ISSUE TITLE
Is there a plan to make liveQuery more compatible with Svelte5 runes

### ISSUE THREAD

**Issue by newholder** [dexie/Dexie.js#2075]
With the introduction of runes, store is kinda obsolete. Is there any plan to support runes once svelte5 released?

I use it like this in svelte5

```svelte
let _friends = liveQuery(async () => {
      return await db.friends
        .where("age")
        .between(18, 65)
        .toArray();
    });

let friends = $derived($_friends)
```

**Comment by dfahlander** [comment]
Thanks for bringing it up. Also asked on stackoverflow: https://stackoverflow.com/questions/78089371/dexie-livequery-with-svelte-5-runes

Need more knowledge about Svelte 5 and runes before I could do anything about it but a PR would be much welcome. Could we introduce some kind of svelte helper library that delivers runes for live queries? A bit like we have for react?

**Comment by molarmanful** [comment]
The recently-added [fromStore](https://svelte-5-preview.vercel.app/docs/imports#svelte-store-fromstore) alongside the wrapper from #1907 worked for me:

```js
import { liveQuery } from 'dexie'
import { fromStore } from 'svelte/store'

export const liveQ = q => fromStore({
  subscribe: (run, invalidate) =>
    liveQuery(q).subscribe(run, invalidate).unsubscribe,
})

// usage

const friends = liveQ(async () => await db.friends.where('age').between(18, 65).toArray())

$inspect(friends.current)
```

**Comment by ericdudley** [comment]
Hello, I am also trying to use Dexie.js in a Svelte5 project with runes. I started by following the existing tutorial [Get started with Dexie in Svelte](https://dexie.org/docs/Tutorial/Svelte) tutorial. However, I got stuck when trying to make a liveQuery reactive to props/state.

I was able to get it working by combining `$derived.by` with a noop reference to the state I want to react to (in this case `namePrefix`). Is this the correct way of parameterizing a live query? Re-creating the liveQuery whenever the parameter changes?

```
<script lang="ts">
	import { db } from '$lib/store';
	import { liveQuery } from 'dexie';

	let { namePrefix }: { namePrefix: string } = $props();

	let friends = $derived.by(() => {
		// noop just to make it reactive
		namePrefix;

		return liveQuery(() => db.friends
				.where('name')
				.startsWithIgnoreCase(namePrefix ?? '')
				.toArray()
		);
	});
</script>
```

**Comment by dfahlander** [comment]
> Hello, I am also trying to use Dexie.js in a Svelte5 project with runes. I started by following the existing tutorial [Get started with Dexie in Svelte](https://dexie.org/docs/Tutorial/Svelte) tutorial. However, I got stuck when trying to make a liveQuery reactive to props/state.
> 
> I was able to get it working by combining `$derived.by` with a noop reference to the state I want to react to (in this case `namePrefix`). Is this the correct way of parameterizing a live query? Re-creating the liveQuery whenever the parameter changes?
> 
> ```
> <script lang="ts">
> 	import { db } from '$lib/store';
> 	import { liveQuery } from 'dexie';
> 
> 	let { namePrefix }: { namePrefix: string } = $props();
> 
> 	let friends = $derived.by(() => {
> 		// noop just to make it reactive
> 		namePrefix;
> 
> 		return liveQuery(() => db.friends
> 				.where('name')
> 				.startsWithIgnoreCase(namePrefix ?? '')
> 				.toArray()
> 		);
> 	});
> </script>
> ```

It's correct to recreate the query whenever a closure it depends on changes. In the background the old query will be unsubscribed and a new query subscribed and the result will stay until the new query emits it's initial value.

**Comment by dfahlander** [comment]
We'd need some svelte-specific library to consume live queries more slick with runes. See also my suggested helper in #2089 - it emits errors and loading status but it still need the `noops` for some reason (or does it). Wouldn't the svelte compiler see the closures passed to the callback? If that's the case, it would be enough with:

```html
<script>
  const friends = $derived(
    liveQuery(
      () => db.friends
        .where('name')
        .startsWith(namePrefix)
        .toArray()
    )
  );
</script>

{#if $friends}
  <ul>
    {#each $friends as friend (friend.id)}
      <li>{friend.name}, {friend.age}</li>
    {/each}
  </ul>
{/if}
```

Could anyone confirm whether the above code would work? It would be more direct than what's suggested so far.


If errors and loading status is requested, then look at how it could be done in #2089. An updated sample using that helper would then be:

```html
<script>
  const friendResult = $derived(
    svelteLiveQuery(
      () => db.friends
        .where('name')
        .startsWith(namePrefix)
        .toArray()
    )
  );
</script>

<!-- isLoading -->
{#if $friendResult.isLoading}
   <p>Loading...</p>
{/if}

<!-- error -->
{#if $friendResult.error}
   <p>Error: $friendResult.error</p>
{/if}

<!-- result -->
{#if $friendResult.value}
<ul>
  {#each $friendResult.value as friend (friend.id)}
    <li>{friend.name}, {friend.age}</li>
  {/each}
</ul>
{/if}


```

**Comment by dusty-phillips** [comment]
I've also been struggling with various incantations of this. It works acceptably in certain toy app scenarios, but it's a frustrating nightmare in a production app.

I think I've got a working solution now, thanks in large part to standing on the shoulders of @braebo, who posted a solution [here](https://github.com/braebo/svelte-5-dexie-test/) (see the "fixed" route). The [QueryRune](https://github.com/braebo/svelte-5-dexie-test/blob/main/src/lib/QueryRune.svelte.ts) works correctly for simple cases.

One gotcha is that it doesn't behave as expected with queries that have derived parameters.So for example, this doesn't work:

```
  const draftQuery = $derived(new QueryRune(liveQuery(() => getSelectedDraft(book?.id))))
```

(`getSelectedDraft` is just a promise that calls into Dexie)

This is not reactive because according to the [Svelte docs](https://svelte.dev/docs/svelte/$derived#Understanding-dependencies):

> Anything read **synchronously** inside the $derived expression (or $derived.by function body) is considered a dependency of the derived state.

(emphasis mine)

Because `liveQuery` is executing **asynchronously**, anything passed into it is not considered reactive. David, I didn't test your `namePrefix` example, but I am pretty confident it would suffer from the same problem.

The least untidy solution I've found is to bring back the concept of ye olde deps array. 😭 I've created a function like this:

```typescript
export function liveRune<T>(
  querier: () => T | Promise<T>,
  ..._dependencies: any[]
): QueryRune<T> | { current: undefined } {
  return new QueryRune(liveQuery(querier))
}
```

I can call it like this:

```typescript
  const draftQuery = $derived(liveRune(() => getSelectedDraft(book?.id), book?.id))
```

The function doesn't do anything with the second `book?.id` but because `$derived` sees it as a synchronous parameter, it treats it reactively things update properly.

I've actually extended this function to short circuit if any of the deps are undefined, so cascading loads can behave sanely when waiting for data from a parent query (book, in this example):

```typescript
export function liveRune<T>(
  querier: () => T | Promise<T>,
  ...dependencies: any[]
): QueryRune<T> | { current: undefined } {
  if (!dependencies.every((x) => x)) {
    return { current: undefined }
  }

  return new QueryRune(liveQuery(querier))
}
```

Svelte's generally done an excellent job of inferring dependencies, so I'm pretty bummed out that it isn't able to do the right thing without an explicit dependency. But it's the best I've found so far.

**Comment by braebo** [comment]
@dusty-phillips thanks for sharing your findings!  I'm not sure if this helps, but I've got derived working reliably [here](https://github.com/braebo/nutab/blob/main/src/lib/data/db.svelte.ts).

**Comment by dusty-phillips** [comment]
> It's correct to recreate the query whenever a closure it depends on changes. In the background the old query will be unsubscribed and a new query subscribed and the result will stay until the new query emits it's initial value.

In practice, recreating the query when dependent state changes is causing the query to very briefly return `undefined` while the new query populates itself. Because the dependent components don't render when the data is undefined, my components momentarily flash out of existence until the defined value returned.

Has anyone else seen/resolved this? I'm experimenting with "store the old value in separate state until the next value resolves", but it makes me nervous.

**Comment by dusty-phillips** [comment]
...the problem is that I also can't sleep when my brain has a puzzle to chew on. This is working for the two toy queries I tried it with:

```typescript
import { untrack } from 'svelte';
import { liveQuery } from 'dexie';

export class DexieQuery<T, A extends any[]> {
	#previous = $state<T>();
	#store = $state<T>();

	current = $derived(this.#store ?? this.#previous);

	constructor(querier: (...args: A) => T | Promise<T>, dependencies: () => A) {
		$effect(() => {
			const deps = dependencies();

			untrack(() => {
				if (this.#store !== undefined) {
					this.#previous = this.#store;
				}
				const store = liveQuery(() => querier(...deps));
				return store.subscribe((value) => {
					this.#store = value;
					this.#previous = undefined;
				}).unsubscribe;
			});
		});
	}
}
```

It's basically a bit of all the previous proposed solutions mixed together. In use it looks like this:

```typescript
	const allFriendsQuery = new DexieQuery(
		() => db.friends.toArray(),
		() => [] as const
	);
	const allFriends = $derived(allFriendsQuery.current);

	const filteredFriendsQuery = new DexieQuery(
		(age) => db.friends.where('age').above(age).toArray(),
		() => [ageFilter] as const
	);
	const filteredFriends = $derived(filteredFriendsQuery.current);

	$inspect('filteredFriends', filteredFriends); // never undefined once set!
```

The `as const` is kind of annoying but otherwise, I think it might work.

**Comment by oliverdowling** [comment]
Is it required to pass the dependencies to the query? This seems to work for me in basic test:

```typescript
export class DexieQuery<T, A extends unknown[]> {
	#previous = $state<T>();
	#store = $state<T>();

	current = $derived(this.#store ?? this.#previous);

	constructor(querier: () => T | Promise<T>, dependencies?: () => A) {
		$effect(() => {
			dependencies?.();

			untrack(() => {
				if (this.#store !== undefined) {
					this.#previous = this.#store;
				}
				const store = liveQuery(() => querier());
				return store.subscribe((value) => {
					this.#store = value;
					this.#previous = undefined;
				}).unsubscribe;
			});
		});
	}
}
// Usage:
const allFriendsQuery = new DexieQuery(async () => await db.friends.toArray());
const filteredFriendsQuery = new DexieQuery(
	async () => await db.friends.where('age').above(age).toArray(),
	() => [age]
);
```

In my use-case, I dont ever change back to `undefined` so I have been using something a bit simpler:

```typescript
export function stateQuery<T>(querier: () => T | Promise<T>, dependencies?: () => unknown[]) {
	const query = $state<{ result?: T }>({});
	$effect(() => {
		dependencies?.();
		return liveQuery(querier).subscribe((result) => {
			if (result !== undefined) {
				query.result = result;
			}
		}).unsubscribe;
	});
	return { query };
}
// Usage:
let { query: friendsQuery } = stateQuery(
	async () => {
		return await db.friends.where('age').above(age).toArray();
	},
	() => [age]
);
```

I also dont bother to use another `$derived` and just reference `friendsQuery.result`

**Comment by dusty-phillips** [comment]
Excellent! To make it even simpler we can return the query directly and not bother with destructuring:

```typescript

export function stateQuery<T>(
	querier: () => T | Promise<T>,
	dependencies: () => unknown[]
): { current?: T } {
	const query = $state<{ current?: T }>({ current: undefined });
	$effect(() => {
		dependencies?.();
		return liveQuery(querier).subscribe((result) => {
			if (result !== undefined) {
				query.current = result;
			}
		}).unsubscribe;
	});
	return query;
}
```

```typescript
	let filteredFriendsQuery = stateQuery(
		async () => {
			return await db.friends.where('age').above(ageFilter).toArray();
		},
		() => [ageFilter]
	);
	const filteredFriends = $derived(filteredFriendsQuery.current);
	$inspect('filteredFriends', filteredFriends); // never undefined once set!
```

**Comment by dfahlander** [comment]
Seems you have found something useful here ! Not sure I grip everything yet, but would it be doable to put this into a library? I would love to update the svelte docs when we have something that seems to be a good solution for svelte runes.

**Comment by dusty-phillips** [comment]
I can try but no promises on timing as I’m pretty swamped! Would you like a PR for a new dexie.js package a la react-live-query, or a third party solution?

**Comment by dfahlander** [comment]
I'm open to whatever would feel comfortable for you - if you'd prefer owning it I'm happy. I you rathe want to hand it over, a recursive copy of libs/dexie-react-hooks would be a good start - remove its react deps, code and tests and put in the deps and code needed for svelte. That would be a nice start as a PR.

**Comment by oliverdowling** [comment]
I had some time today so I've made a pull request, let me know if there are any issues you might like me to address.

**Comment by dlintw** [comment]
I tried to use the workaround solution, please advise me better solution. source code in.

https://github.com/dlintw/svelte5-ts-tailwind-go

**Comment by dnotes** [comment]
Hey, I think I was a little confused when I ran across this issue, but I've since discovered that Svelte stores are NOT deprecated or necessarily obsolete; they are just for a particular set of circumstances. The [documentation](https://svelte.dev/docs/svelte/stores) implies this, and there are [discussion threads](https://github.com/sveltejs/svelte/discussions/15885) that seem to make it pretty clear. I'm not sure if this really matches one of the situations where you'd *need* to use Stores instead of Runes, but they still work.

In short, even when we get a Dexie/Runes solution working, please don't throw out the existing Stores solution. Also if you've arrived here because you too were confused about Svelte stores being outdated, they still work and there are no plans to get rid of them, so you don't have to wait until this is ready to use Dexie with Svelte 5.

**Comment by MrVauxs** [comment]
For me the primary problem I am coming across is that you cannot easily change the query function and have to rebuild it. Really looking forward to seeing the library addressing this problem.

If its any consideration (though likely out of scope of this issue since this is about reworking how liveQuery works), I would please ask to look at [runed's resource](https://runed.dev/docs/utilities/resource) utility, primarily the fact that it includes `loading` and `error` reactive getters, allowing for more than just plain `undefined` that comes out of current liveQuery.

On that above note, thanks to [philips code above](https://github.com/dexie/Dexie.js/issues/2075#issuecomment-2573032613) I was able to reintroduce those states. Unfortunately it cannot tell what kind of error has been made, but one step at a time.

```ts
interface queryResult<T> { current?: T; loading: boolean; failed: boolean }
function stateQuery<T>(
	querier: () => T | Promise<T>,
	dependencies: () => unknown[],
): queryResult<T> {
	const query = $state<queryResult<T>>({ current: undefined, loading: false, failed: false });
	$effect(() => {
		dependencies?.();
		query.loading = true;
		return liveQuery(querier).subscribe((result) => {
			if (result !== undefined) {
				query.current = result;
				query.failed = false;
			}
			else {
				query.failed = true;
			}
			query.loading = false;
		}).unsubscribe;
	});
	return query;
}
```

This allows having knowledge of whether a given query has failed or is being awaited while still keeping the previous result.

**Comment by dfahlander** [comment]
Library is now on npm: https://www.npmjs.com/package/dexie-svelte-query
